kong/plugins/ldap-auth/asn1.lua

```diff
@@ -27,183 +53,169 @@ _M.BERCLASS = {
-      return self:decodeOID(encStr, elen, pos)
+    local ret = C.ASN1_get_object(strpp, lenp, tagp, classp, stop - start + 1)
+    if band(ret, 0x80) == 0x80 then
+        error("der with error encoding", ret)
     end
 
-    -- Context specific tags
-    self.decoder["30"] = function(self, encStr, elen, pos)
-      return self:decodeSeq(encStr, elen, pos)
+    local constructed = false
+    if band(ret, 0x20) == 0x20 then
+        constructed = true
     end
-  end,
 
-  registerTagDecoders = function(self, tagDecoders)
-    self:registerBaseDecoders()
-
-    for k, v in pairs(tagDecoders) do
-      self.decoder[k] = v
-    end
-  end,
+    return tagp[0], classp[0], tonumber(lenp[0]), strpp[0] - s_der, constructed
+  end
+end
+_M.asn1_get_object = asn1_get_object
 
-  decodeLength = function(encStr, pos)
-    local elen
 
-    pos, elen = bunpack(encStr, "C", pos)
-    if elen > 128 then
-      elen = elen - 128
-      local elenCalc = 0
-      local elenNext
+--[[
+Encoded LDAP Result: https://ldap.com/ldapv3-wire-protocol-reference-ldap-result/
 
-      for i = 1, elen do
-        elenCalc = elenCalc * 256
-        pos, elenNext = bunpack(encStr, "C", pos)
-        elenCalc = elenCalc + elenNext
-      end
+30 0c -- Begin the LDAPMessage sequence
+   02 01 03 -- The message ID (integer value 3)
+   69 07 -- Begin the add response protocol op
+      0a 01 00 -- success result code (enumerated value 0)
+      04 00 -- No matched DN (0-byte octet string)
+      04 00 -- No diagnostic message (0-byte octet string)
+--]]
 
-      elen = elenCalc
-    end
+local function parse_ldap_result(der)
+  local p = ffi_cast("const unsigned char *", der)
+  cucharpp[0] = p
+  local _, _, len, offset = asn1_get_object(der)
 
-    return pos, elen
-  end,
+  -- message ID (integer)
+  local asn1_int = C.d2i_ASN1_INTEGER(nil, cucharpp, #der)
+  local id = C.ASN1_INTEGER_get(asn1_int)
 
-  decodeSeq = function(self, encStr, len, pos)
-    local seq = {}
-    local sPos = 1
-    local sStr
+  _, _, len, offset = asn1_get_object(der, offset + len)
+  local _, op_int = bunpack(der, "C", offset - 1)
 
-    pos, sStr = bunpack(encStr, "A" .. len, pos)
+  local op = _M.intToBER(op_int)
 
-    while (sPos < len) do
-      local newSeq
+  -- success result code
+  cucharpp[0] = p + offset
+  asn1_int = C.d2i_ASN1_ENUMERATED(nil, cucharpp, len)
+  local res = C.ASN1_ENUMERATED_get(asn1_int)
 
-      sPos, newSeq = self:decode(sStr, sPos)
-      if not newSeq and self.stoponerror then
-        break
-      end
+  -- No matched DN (octet string)
+  local asn1_str = C.d2i_ASN1_OCTET_STRING(nil, cucharpp, #der)
+  local err1 = C.ASN1_STRING_get0_data(asn1_str)
 
-      insert(seq, newSeq)
-    end
+  -- No diagnostic message (octet string)
+  asn1_str = C.d2i_ASN1_OCTET_STRING(nil, cucharpp, #der)
+  local err2 = C.ASN1_STRING_get0_data(asn1_str)
 
-    return pos, seq
-  end,
+  -- free
+  ffi.gc(asn1_str, C.ASN1_STRING_free)
+  ffi.gc(asn1_int, C.ASN1_INTEGER_free)
```

Comment:
typically we only need to set ffi.gc if the cdata is returned or being passed along and we don't know when it's not needed. for here we can just do 

```suggestion
  C.ASN1_STRING_free(asn1_str)
  C.ASN1_INTEGER_free(asn1_int)
```

Reply:
Thanks!